django/utils/datastructures.py

```diff
@@ -308,9 +308,11 @@ class CaseInsensitiveMapping(Mapping):
     """
 
     def __init__(self, data):
-        if not isinstance(data, Mapping):
-            data = {k: v for k, v in _destruct_iterable_mapping_values(data)}
-        self._store = {k.lower(): (k, v) for k, v in data.items()}
+        if isinstance(data, Mapping):
+            data_items = data.items()
+        else:
+            data_items = _destruct_iterable_mapping_values(data)
```

Comment:
There are only two uses of `_destruct_iterable_mapping_values()` and we use this same pattern exactly.

I think that you could push the `isinstance(..., Mapping)` check into that function.

Reply:
As a separate commit, I think that it is also probably worth moving that global function to be a `@staticmethod` on the `CaseInsensitiveMapping` class. It is closely associated with this class and would avoid the need to import it separately in `django.http.response`.

Reply:
> There are only two uses of `_destruct_iterable_mapping_values()` and we use this same pattern exactly.
> 
> I think that you could push the `isinstance(..., Mapping)` check into that function.

`_destruct_iterable_mapping_values` is a generator function, moving the condition to it will result in adding an additional `for`-loop to iterate over mappings